service-base/src/main/java/org/eclipse/hono/service/tenant/BaseTenantService.java

```diff
@@ -162,15 +162,27 @@ public final void processTenantMessage(final Message<JsonObject> tenantMsg) {
                 remove(tenantId, result -> reply(tenantMsg, result));
                 break;
             default:
-                log.debug("invalid action in request message [{}]", subject);
-                reply(tenantMsg, TenantResult.from(HTTP_BAD_REQUEST));
+                processCustomTenantMessage(tenantMsg, subject);
+                break;
             }
         } catch (final ClassCastException e) {
             log.debug("malformed request message [{}]", e.getMessage());
             reply(tenantMsg, TenantResult.from(HTTP_BAD_REQUEST));
         }
     }
 
+    /**
+     * Override the following method to extend TenantService with implementation
```

Comment:
IMHO this comment should explain what the purpose of this method is and when and in what context it will be invoked so that implementers can better understand what they need to do. In particular, the comment should include instructions on how to actually convey a response back to the client.

Same applies to the other two services.